HiSkim/HiOnia2MuMu/src/HiOnia2MuMuPAT.cc

```diff
@@ -520,31 +563,400 @@ HiOnia2MuMuPAT::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+
+    	    	    TransientTrack tt = theTTBuilder->build(track);
+    	    	    pair<bool,Measurement1D> tkPVdist = IPTools::absoluteImpactParameter3D(tt,theOriginalPV);
+
+    	    	    if (!tkPVdist.first) continue;
+    	    	    if (tkPVdist.second.significance()>3) continue;
+    	    	    if (track.ptError()/track.pt()>0.1) continue;
+
+    	    	    // do not count the three muons
+    	    	    if (rmu1 != 0 && rmu1->innerTrack().key() == itVtx->key()) continue;
+    	    	    if (rmu2 != 0 && rmu2->innerTrack().key() == itVtx->key()) continue;
+    	    	    if (rmu3 != 0 && rmu3->innerTrack().key() == itVtx->key()) continue;
+
+    	    	    vertexWeight += theOriginalPV.trackWeight(*itVtx);
+    	    	    if(theOriginalPV.trackWeight(*itVtx) > 0.5){
+    	    	      countTksOfPV++;
+    	    	      sumPTPV += track.pt();
+    	    	    }
+    	    	  }
+    	      } catch (std::exception & err) {std::cout << " Counting tracks from PV, fails! " << std::endl; return ; }
+    	    }
+    	    userBcInt["countTksOfPV"] = countTksOfPV;
+    	    userBcFloat["vertexWeight"] = (float) vertexWeight;
+    	    userBcFloat["sumPTPV"] = (float) sumPTPV;
+         
+    	    if (addMuonlessPrimaryVertex_) {
+    	      userBcVertex["muonlessPV"] = thePrimaryV;
+    	      userBcVertex["PVwithmuons"] = theOriginalPV;
+    	    } else {
+    	      userBcVertex["PVwithmuons"] = thePrimaryV;
+    	    }
+
+    	    // lifetime using PV
+    	    pvtx.SetXYZ(thePrimaryV.position().x(),thePrimaryV.position().y(),0);
+    	    TVector3 vdiff = vtx - pvtx;
+    	    double cosAlpha = vdiff.Dot(pperp)/(vdiff.Perp()*pperp.Perp());
+    	    Measurement1D distXY = vdistXY.distance(Vertex(TrimuVertex), thePrimaryV);
+    	    double ctauPV = distXY.value()*cosAlpha*6.276/pperp.Perp();
+    	    GlobalError v1e = (Vertex(TrimuVertex)).error();
+    	    GlobalError v2e = thePrimaryV.error();
+    	    AlgebraicSymMatrix33 vXYe = v1e.matrix()+ v2e.matrix();
+    	    double ctauErrPV = sqrt(ROOT::Math::Similarity(vpperp,vXYe))*6.276/(pperp.Perp2());
+
+    	    userBcFloat["ppdlPV"] = ctauPV;
+    	    userBcFloat["ppdlErrPV"] = ctauErrPV;
+    	    userBcFloat["cosAlpha"] = cosAlpha;
+
+    	    pvtx3D.SetXYZ(thePrimaryV.position().x(),thePrimaryV.position().y(),thePrimaryV.position().z());
+    	    TVector3 vdiff3D = vtx3D - pvtx3D;
+    	    double cosAlpha3D = vdiff3D.Dot(pxyz)/(vdiff3D.Mag()*pxyz.Mag());
+    	    Measurement1D distXYZ = vdistXYZ.distance(Vertex(TrimuVertex), thePrimaryV);
+    	    double ctauPV3D = distXYZ.value()*cosAlpha3D*6.276/pxyz.Mag();
+    	    double ctauErrPV3D = sqrt(ROOT::Math::Similarity(vpxyz,vXYe))*6.276/(pxyz.Mag2());
+          
+    	    userBcFloat["ppdlPV3D"] = ctauPV3D;
+    	    userBcFloat["ppdlErrPV3D"] = ctauErrPV3D;
+    	    userBcFloat["cosAlpha3D"] = cosAlpha3D;
+
+	    // if (addMuonlessPrimaryVertex_) {
+	    //   // lifetime using Original PV
```

Comment:
Please remove all the commented lines below

Reply:
I would like to keep this, in case I ever want the lifetime with beamspot etc. for the Bc/trimuon.